translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+				// has a different Resource.
+				separateRS := &data.ResourceSpans{}
+				separateRS.SetResource(ocNodeResourceToInternal(nil, ocSpan.Resource))
+				separateRS.SetSpans([]*data.Span{dest})
+				resourceSpanList = append(resourceSpanList, separateRS)
+			} else {
+				// Otherwise add the span to the first ResourceSpans item.
+				ptrs = append(ptrs, dest)
+			}
+		}
+
+		resourceSpans.SetSpans(ptrs)
+	}
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) data.TimestampUnixNano {
+	return data.TimestampUnixNano(uint64(internal.TimestampToTime(ts).UnixNano()))
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	dest.SetTraceID(data.TraceIDFromBytes(src.TraceId))
+	dest.SetSpanID(data.SpanIDFromBytes(src.SpanId))
+	dest.SetTraceState(ocTraceStateToInternal(src.Tracestate))
+	dest.SetParentSpanID(data.SpanIDFromBytes(src.ParentSpanId))
+	dest.SetName(truncableStringToStr(src.Name))
+	dest.SetKind(ocSpanKindToInternal(src.Kind, src.Attributes))
+	dest.SetStartTime(timestampToUnixnano(src.StartTime))
+	dest.SetEndTime(timestampToUnixnano(src.EndTime))
+	dest.SetAttributes(ocAttrsToInternal(src.Attributes))
+	dest.SetEvents(events)
+	dest.SetDroppedEventsCount(droppedEventCount)
+	dest.SetLinks(links)
+	dest.SetDroppedLinksCount(droppedLinkCount)
+	dest.SetStatus(ocStatusToInternal(src.Status))
+}
+
+func ocStatusToInternal(ocStatus *octrace.Status) data.SpanStatus {
+	if ocStatus == nil {
+		return data.SpanStatus{}
+	}
+	return data.NewSpanStatus(data.StatusCode(ocStatus.Code), ocStatus.Message)
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToInternal(ocTracestate *octrace.Span_Tracestate) data.TraceState {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return data.TraceState(sb.String())
+}
+
+func ocAttrsToInternal(ocAttrs *octrace.Span_Attributes) data.Attributes {
+	if ocAttrs == nil {
+		return data.Attributes{}
+	}
+
+	attrCount := len(ocAttrs.AttributeMap)
+
+	attrMap := make(data.AttributesMap, attrCount)
+	values := data.NewAttributeValueSlice(attrCount)
+	i := 0
+	for key, ocAttr := range ocAttrs.AttributeMap {
+		switch attribValue := ocAttr.Value.(type) {
+		case *octrace.AttributeValue_StringValue:
+			values[i].SetString(truncableStringToStr(attribValue.StringValue))
+
+		case *octrace.AttributeValue_IntValue:
+			values[i].SetInt(attribValue.IntValue)
+
+		case *octrace.AttributeValue_BoolValue:
+			values[i].SetBool(attribValue.BoolValue)
+
+		case *octrace.AttributeValue_DoubleValue:
+			values[i].SetDouble(attribValue.DoubleValue)
+
+		default:
+			str := "<Unknown OpenCensus Attribute>"
```

Comment:
nit: Should probably say something like `"<Unknown OpenCensus attribute value type>"`

Reply:
done.